mne/rank.py

```diff
@@ -211,24 +226,165 @@ def _get_rank_sss(inst):
+        To get the total rank, you can use ``sum(rank.values())``.
+
+    Notes
+    -----
+    The ``rank`` parameter can be:
+
+    :class:`python:int`
+        This value is used as the MEG rank. For other channel types,
+        rank is taken from ``info``.
+    :data:`python:None`
+        the rank will be automatically estimated
+        from the data after proper scaling of the different
+        channel types.
+    ``'info'`` (default)
+        Rank is inferred from `info`. If data have been processed
+        with Maxwell filtering, the Maxwell filtering header is used.
+        Otherwise, the channel counts themselves are used.
+        In both cases, the number of projectors is accounted for.
+    ``'full'``
+        Rank is assumed to be full, i.e. equal to the
+        number of good channels. If a `Covariance`
+        is passed, this can make sense if it has been (possibly
+        improperly) regularized without taking into account the true
+        data rank.
+
+    .. versionadded:: 0.18
+    """
+    from .io.base import BaseRaw
+    from .epochs import BaseEpochs
+    from . import Covariance
+
+    rank = _check_rank(rank)
+
+    if isinstance(inst, Covariance):
+        if info is None:
+            raise ValueError('info cannot be None if inst is a Covariance.')
+        inst = pick_channels_cov(
+            inst, set(inst['names']) & set(info['ch_names']))
+        info = pick_info(info, [info['ch_names'].index(name)
+                                for name in inst['names']])
     else:
-        raise ValueError('There is no `max_info` here. Sorry.')
-    return _get_sss_rank(max_info)
+        info = inst.info
+
+    if isinstance(rank, str):  # string
+        rank_type = 'info'
+        info_type = rank
+        rank = dict()
+    else:  # None, dict, or int
+        rank_type = 'estimated'
+        if not isinstance(rank, dict):  # dict is pass-through
+            if rank is not None:  # int
+                rank = dict(meg=int(operator.index(rank)))
+            else:  # None
+                rank = dict()
+    assert isinstance(rank, dict)  # should be guaranteed by _check_rank
+
+    picks_list = _picks_by_type(info, meg_combined=True, ref_meg=False,
+                                exclude='bads')
+    for ch_type, picks in picks_list:
+        if ch_type in rank:
+            continue
+        ch_names = [info['ch_names'][pick] for pick in picks]
+        if rank_type == 'info':
+            # use info
+            rank[ch_type] = _info_rank(info, ch_type, picks, info_type)
+            if rank != 'full':
+                rank[ch_type] -= make_projector(info['projs'], ch_names)[1]
+        else:
+            # Use empirical estimation
+            assert rank_type == 'estimated'
+            if isinstance(inst, BaseRaw):
+                reject_by_annotation = 'omit'
+                data = inst.get_data(picks, None, None, reject_by_annotation)
+                rank[ch_type] = estimate_rank(data)
```

Comment:
this should be using _estimate_rank_meeg_signals and should take the scalings as param.

I would even suggest to deprecate the estimate_rank function but this can be done later.